OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnHealth.cs

```diff
@@ -0,0 +1,99 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Applies a condition to the actor at when its health is between 2 specific values.")]
+	public class GrantConditionOnHealthInfo : ITraitInfo, IRulesetLoaded, Requires<IHealthInfo>
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("Condition to grant.")]
+		public readonly string Condition = null;
+
+		[Desc("Play a random sound from this list when enabled.")]
+		public readonly string[] EnabledSounds = { };
+
+		[Desc("Play a random sound from this list when disabled.")]
+		public readonly string[] DisabledSounds = { };
+
+		[Desc("Minimum level of health at which to grant the condition.")]
+		public readonly int MinHP = 0;
+
+		[Desc("Maximum level of health at which to grant the condition.")]
+		public readonly int MaxHP = 0;
+
+		[Desc("Is the condition irrevokable once it has been granted?")]
+		public readonly bool GrantPermanently = false;
+
+		public object Create(ActorInitializer init) { return new GrantConditionOnHealth(init.Self, this); }
+
+		public void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			var health = ai.TraitInfo<IHealthInfo>();
+			if (health.MaxHP < MinHP)
+				throw new YamlException("Minimum HP ({0}) for GrantConditionOnHealth can't be more than actor's Maximum HP ({1})".F(MinHP, health.MaxHP));
+		}
+	}
+
+	public class GrantConditionOnHealth : INotifyCreated, INotifyDamage
+	{
+		readonly GrantConditionOnHealthInfo info;
+		readonly IHealth health;
+		readonly int maxHP;
+
+		ConditionManager conditionManager;
+		int conditionToken = ConditionManager.InvalidConditionToken;
+
+		public GrantConditionOnHealth(Actor self, GrantConditionOnHealthInfo info)
+		{
+			this.info = info;
+			health = self.Trait<IHealth>();
+			maxHP = info.MaxHP > 0 ? info.MaxHP : health.MaxHP;
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			conditionManager = self.TraitOrDefault<ConditionManager>();
```

Comment:
Please use just `.Trait<ConditionManager>()`. One shouldn't use this trait without one anyway, but might get an NRE with the current code (you don't check for `null` before the call below).

Reply:
All other traits have proper null checks, i guess it would be better add checks than use .Trait<>.

Reply:
Ok, then add the check.

Reply:
This is the case in GrantConditionOnDamageState too, should i update it too?

Reply:
Huh, I thought you were going to add the checks. But if you use `Trait` instead of `TraitOrDefault` you should remove the `null` checks.